sdk/core/System.ClientModel/src/Pipeline/HttpClientPipelineTransport.Response.cs

```diff
@@ -39,17 +46,93 @@ protected override PipelineResponseHeaders GetHeadersCore()
 
         public override Stream? ContentStream
         {
-            get => _contentStream;
-            set
+            get
             {
-                // Make sure we don't dispose the content if the stream was replaced
-                _httpResponse.Content = null;
+                if (_contentStream is not null)
+                {
+                    return _contentStream;
+                }
+
+                if (_bufferedContent is not null)
+                {
+                    return _bufferedContent.ToStream();
+                }
 
+                return null;
+            }
+            set
+            {
                 _contentStream = value;
+
+                // Invalidate the cache since the source-stream has been replaced.
+                _bufferedContent = null;
             }
         }
 
-        #region IDisposable
+        public override BinaryData Content
+        {
+            get
+            {
+                if (_bufferedContent is not null)
+                {
+                    return _bufferedContent;
+                }
+
+                if (_contentStream is null || _contentStream is MemoryStream)
+                {
+                    ReadContent();
+                }
+
+                throw new InvalidOperationException($"The response is not buffered.");
+            }
+        }
+
+        public override BinaryData ReadContent(CancellationToken cancellationToken = default)
+            => ReadContentSyncOrAsync(cancellationToken, async: false).EnsureCompleted();
+
+        public override async ValueTask<BinaryData> ReadContentAsync(CancellationToken cancellationToken = default)
+            => await ReadContentSyncOrAsync(cancellationToken, async: true).ConfigureAwait(false);
+
+        private async ValueTask<BinaryData> ReadContentSyncOrAsync(CancellationToken cancellationToken, bool async)
+        {
+            if (_bufferedContent is not null)
+            {
+                // Content has already been buffered.
+                return _bufferedContent;
+            }
+
+            if (_contentStream == null)
+            {
+                // Content is not buffered but there is no source stream.
+                // Our contract from Azure.Core is to return BinaryData.Empty in this case.
+                _bufferedContent = s_EmptyBinaryData;
+                return _bufferedContent;
+            }
+
+            // ContentStream still holds the source stream.  Buffer the content
+            // and dispose the source stream.
+            BufferedContentStream bufferStream = new();
+
+            if (async)
+            {
+                await _contentStream.CopyToAsync(bufferStream, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                _contentStream.CopyTo(bufferStream, cancellationToken);
+            }
+
+            _contentStream.Dispose();
```

Comment:
it would be good to call DisposeAsync in async mode.

Reply:
Good suggestion

Reply:
This is also not available on netstandard2.0.  Depending on how strongly you feel about this, we can polyfil it, but I'd prefer not to block the PR on it.

Reply:
Good suggestion - I will do this